packages/flutter_tools/lib/src/version.dart

```diff
@@ -42,18 +35,47 @@ class FlutterVersion {
 
     _frameworkRevision = _runGit('git log -n 1 --pretty=format:%H');
     _frameworkAge = _runGit('git log -n 1 --pretty=format:%ar');
+
+    final String version = _runGit('git describe --match v*.*.* --exclude *-* --first-parent --long --tags');
+    final RegExp versionPattern = new RegExp('^v([0-9]+)\.([0-9]+)\.([0-9]+\)-([0-9]+)-g([a-f0-9]+)\$');
+    final List<String> parts = versionPattern.matchAsPrefix(version)?.groups(<int>[1, 2, 3, 4, 5]);
+    if (parts != null && _frameworkRevision.startsWith(parts[4])) {
+      if (parts[3] == '0') {
+        _frameworkVersion = '${parts[0]}.${parts[1]}.${parts[2]}';
+      } else {
+        _frameworkVersion = '${parts[0]}.${parts[1]}.${int.parse(parts[2]) + 1}-pre.${parts[3]}';
+      }
+    } else {
+      printTrace('Could not interpret results of "git describe": $version');
+      _frameworkVersion = 'unknown';
+    }
   }
 
   final Clock _clock;
 
   String _repositoryUrl;
   String get repositoryUrl => _repositoryUrl;
 
+  static Set<String> officialChannels = new Set<String>.from(<String>[
+    'master',
+    'dev',
+    'beta',
+    'release',
+  ]);
+
+  static Map<String, String> obsoleteBranches = <String, String>{
```

Comment:
Add docs explaining what this maps to and from

Reply:
done.